### PR TITLE
Deprecate dot-prefixed indices and composable template index patterns

### PR DESCRIPTION
**Pull request by dakrone**
This commit adds a module emitting a deprecation warning when a dot-prefixed index is manually or automatically created, or when a composable index template with an index pattern that uses a dot-prefix is created. This pattern warns that in the future these indices will not be allowed. In a future breaking change (10.0.0 maybe?) the deprecation can then be changed to an exception.

These deprecations are only displayed when a non-operator user is using the API (one that does not set the `X-elastic-product-origin` header).

### CODE REVIEW

**Comment by dakrone** [comment]
Related to https://github.com/elastic/elasticsearch/pull/112623

File: modules/dot-prefix-validation/build.gradle [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  classname 'org.elasticsearch.validation.DotPrefixValidationPlugin'
+}
+
+restResources {
+  restApi {
+    include '_common', 'indices', 'index', 'cluster', 'nodes', 'get', 'ingest', 'bulk', 'reindex'
+  }
+}
+
+tasks.named("yamlRestTestV7CompatTest").configure { enabled = false };
```
Comment (mark-vieira) [inline modules/dot-prefix-validation/build.gradle:23]: This task no longer exists in `main`.

**Review by mark-vieira — CHANGES_REQUESTED** [review]

File: modules/dot-prefix-validation/build.gradle (hunk 1 above)
Comment (dakrone) [reply modules/dot-prefix-validation/build.gradle]: Yep, just removed that one since it caused a bunch of failures.

**Review by mark-vieira — APPROVED** [review]

File: modules/dot-prefix-validation/src/yamlRestTest/resources/rest-api-spec/test/dot_prefix/10_basic.yml [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            index: original
+          dest:
+            index: .reindex2
+
+  - do:
+      ingest.delete_pipeline:
+        id: mypipeline
+
+---
+"Reject index template that has a dot prefix index pattern":
```
Comment (gmarouli) [inline modules/dot-prefix-validation/src/yamlRestTest/resources/rest-api-spec/test/dot_prefix/10_basic.yml:165]: Same here

**Review by gmarouli — APPROVED** [review]
LGTM @dakrone , thanks for addressing this. I left some minor comments 🚀

File: modules/dot-prefix-validation/src/yamlRestTest/resources/rest-api-spec/test/dot_prefix/10_basic.yml [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  - do:
+      headers: { X-elastic-product-origin: kibana }
+      indices.create:
+        index: <.myindex2-{now/d}>
+        body:
+          settings:
+            index.hidden: true
+
+---
+"Reject auto-creation of dot-prefixed indices":
```
[the remark this answers is not in the document]
Comment (dakrone) [reply modules/dot-prefix-validation/src/yamlRestTest/resources/rest-api-spec/test/dot_prefix/10_basic.yml:47]: Yep, I took these tests from the other version of this, I'll adjust the names.

File: modules/dot-prefix-validation/src/main/java/org/elasticsearch/validation/DotPrefixValidator.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.tasks.Task;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public abstract class DotPrefixValidator<RequestType> implements MappedActionFilter {
```
[the remark this answers is not in the document]
Comment (dakrone) [reply modules/dot-prefix-validation/src/main/java/org/elasticsearch/validation/DotPrefixValidator.java:29]: Certainly, I will add some

**Pull request merged**